src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs

```diff
@@ -43,6 +43,11 @@ public class SessionStateTempDataProvider : ITempDataProvider
                 // If we got it from Session, remove it so that no other request gets it
                 session.Remove(TempDataSessionStateKey);
             }
+            else
+            {
+                // Initializing empty session so that it is established before the headers are sent.
```

Comment:
Add to this comment - we're calling `Save()` after the response has been sent which is why need the placeholder.